lib/spack/spack/environment.py

```diff
@@ -760,15 +764,18 @@ def _set_user_specs_from_lockfile(self):
             )
         }
 
-    def clear(self):
+    def clear(self, re_read=False):
         self.spec_lists = {user_speclist_name: SpecList()}  # specs from yaml
         self.dev_specs = {}               # dev-build specs from yaml
         self.concretized_user_specs = []  # user specs from last concretize
         self.concretized_order = []       # roots of last concretize, in order
         self.specs_by_hash = {}           # concretized specs by hash
-        self.new_specs = []               # write packages for these on write()
         self._repo = None                 # RepoPath for this env (memoized)
         self._previous_active = None      # previously active environment
+        if not re_read:
+            # things that cannot be recreated from file
+            self.new_specs = []               # write packages for these on write()
+            self.new_installs = []            # write modules for these on write()
```

Comment:
It's weird to me that we only write modules for brand new installs.  I think this makes environments nondeterministic -- people will be irritated that their modules are not regenerated.

Specifically, if I:
1. Install a bunch of stuff, or have, say, an upstream with a bunch of preexisting installations
2. Make an env with its own `env-modules:` module set and some custom settings
3. `spack install` the env

I would expect that the `env-modules` set would be generated.  It looks like it's not -- which means that we'll have weird reproducibility issues with env depending on what is already installed in a spack instance.

Is this an accurate description of the current behavior?

Reply:
Yes, that's an accurate description of the current behavior, and it's not changed in this PR.